node/runtime/src/impls.rs

```diff
@@ -0,0 +1,171 @@
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Some configurable implementations as associated type for the substrate runtime.
+
+use node_primitives::Balance;
+use runtime_primitives::weights::{Weight, FeeMultiplier, MAX_TRANSACTIONS_WEIGHT, IDEAL_TRANSACTIONS_WEIGHT};
+use runtime_primitives::traits::{Convert, Zero};
+use crate::{Perbill, Balances};
+
+/// Struct that handles the conversion of Balance -> `u64`. This is used for staking's election
+/// calculation.
+pub struct CurrencyToVoteHandler;
+
+impl CurrencyToVoteHandler {
+	fn factor() -> Balance { (Balances::total_issuance() / u64::max_value() as Balance).max(1) }
+}
+
+impl Convert<Balance, u64> for CurrencyToVoteHandler {
+	fn convert(x: Balance) -> u64 { (x / Self::factor()) as u64 }
+}
+
+impl Convert<u128, Balance> for CurrencyToVoteHandler {
+	fn convert(x: u128) -> Balance { x * Self::factor() }
+}
+
+/// A struct that updates the fee multiplier based on the saturation level of the previous block.
+/// This should typically be called per-block.
+///
+/// This assumes that weight is a numeric value in the u32 range.
+///
+/// Formula:
+///   diff = (ideal_weight - current_block_weight)
+///   v = 0.00004
+///   fee = weight * (1 + (v . diff) + v^2 (v . d)^2 / 2)
+///
+/// https://research.web3.foundation/en/latest/polkadot/Token%20Economics/#relay-chain-transaction-fees
+pub struct FeeMultiplierUpdateHandler;
+impl Convert<Weight, FeeMultiplier> for FeeMultiplierUpdateHandler {
+	fn convert(block_weight: Weight) -> FeeMultiplier {
+		let max_fraction = 1_000_000_000_u128;
+		let ideal = IDEAL_TRANSACTIONS_WEIGHT as u128;
+		let max = MAX_TRANSACTIONS_WEIGHT as u128;
+		let block_weight = block_weight as u128;
+
+		let from_max_to_per_fraction = |x: u128| {
+			if let Some(x_fraction) = x.checked_mul(max_fraction) {
+				x_fraction / max
+			} else {
+				max_fraction
+			}
+		};
+
+		let collapse_mul = |a: u128, b| {
+			if let Some(v) = a.checked_mul(b) {
+				v
+			} else {
+				// collapse to zero if it overflow. For now we don't have the accuracy needed to
+				// compute this trivially with u128.
+				Zero::zero()
+			}
+		};
+
+		// determines if the first_term is positive
+		let positive = block_weight >= ideal;
+		let diff = block_weight.max(ideal) - block_weight.min(ideal);
+
+		// 0.00004 = 4/100_000 = 40_000/10^9
+		let v = 40_000;
+		// 0.00004^2 = 16/10^10 ~= 2/10^9
+		let v_squared = 2;
+
+		let mut first_term = v;
+		first_term = collapse_mul(first_term, from_max_to_per_fraction(diff));
+		first_term = first_term / max_fraction;
+
+		// It is very unlikely that this will exist (in our poor perbill estimate) but we are giving
+		// it a shot.
+		let mut second_term = v_squared;
+		second_term = collapse_mul(second_term, from_max_to_per_fraction(diff));
+		second_term = collapse_mul(second_term, from_max_to_per_fraction(diff) / 2);
+		second_term = second_term / max_fraction;
+		second_term = second_term / max_fraction;
```

Comment:
it is correct but as a note we should be able to implement `multiply_by_rational_u64(value: u64, num: u64, den: u64)` I think

and then do `second_term=multiply_rational_u64(vsquared, (diff as u64).pow(2), (max as u64).pow(2))/2`